tests/xrt/56_xclbin/main.cpp

```diff
@@ -0,0 +1,236 @@
+  auto bo0_map = bo0.map<int*>();
+  auto bo1_map = bo1.map<int*>();
+  std::fill(bo0_map, bo0_map + COUNT, 0);
+  std::fill(bo1_map, bo1_map + COUNT, 0);
+
+  // Fill our data sets with pattern
+  int foo = 0x10;
+  int bufReference[COUNT];
+  for (int i = 0; i < COUNT; ++i) {
+    bo0_map[i] = 0;
+    bo1_map[i] = i;
+    bufReference[i] = i + i * foo;
+  }
+
+  bo0.sync(XCL_BO_SYNC_BO_TO_DEVICE, DATA_SIZE, 0);
+  bo1.sync(XCL_BO_SYNC_BO_TO_DEVICE, DATA_SIZE, 0);
+
+  auto run = simple(bo0, bo1, 0x10);
+  run.wait();
+
+  //Get the output;
+  std::cout << "Get the output data from the device" << std::endl;
+  bo0.sync(XCL_BO_SYNC_BO_FROM_DEVICE, DATA_SIZE, 0);
+
+  // Validate our results
+  if (std::memcmp(bo0_map, bufReference, DATA_SIZE))
+    throw std::runtime_error("Value read back does not match reference");
+
+  return 0;
+}
+
+int
+run(int argc, char** argv)
+{
+  if (argc < 3) {
+    usage();
+    return 1;
+  }
+
+  std::string xclbin_fnm;
+  bool verbose = false;
+  unsigned int device_index = 0;
+
+  std::vector<std::string> args(argv+1,argv+argc);
+  std::string cur;
+  for (auto& arg : args) {
+    if (arg == "-h") {
+      usage();
+      return 1;
+    }
+    else if (arg == "-v") {
+      verbose = true;
+      continue;
+    }
+
+    if (arg[0] == '-') {
+      cur = arg;
+      continue;
+    }
+
+    if (cur == "-k")
+      xclbin_fnm = arg;
+    else if (cur == "-d")
+      device_index = std::stoi(arg);
+    else
+      throw std::runtime_error("Unknown option value " + cur + " " + arg);
+  }
+
+  if (xclbin_fnm.empty())
+    throw std::runtime_error("FAILED_TEST\nNo xclbin specified");
+
+  if (device_index >= xclProbe())
+    throw std::runtime_error("Cannot find device index (" + std::to_string(device_index) + ") specified");
+
+
+  auto xclbin = xrt::xclbin(xclbin_fnm); // C++ API, construct xclbin object from filename
+  
+  std::cout << "Targeting xsa platform: " << xclbin.get_xsa_name() << std::endl;
+
+  std::cout << "Xclbin has uuid: " << xclbin.get_uuid() << std::endl;
+
+  std::vector<std::string> cuNames = xclbin.get_cu_names();
+ 
+  for (auto& cuName : cuNames)
+    std::cout << cuName << " ";
+  std::cout << std::endl;
+
+  if (cuNames[0] != "simple:simple_1")
+    throw std::runtime_error("FAILED_TEST\nCould not read correct kernel name, expected: simple:simple_1");
+
+  std::vector<char> data = xclbin.get_data();
+  for(int i=0;i<7;++i)
+    std::cout << data[i] << " ";
+  std::cout << std::endl;
+
+  auto xclbinHandle = xrtXclbinAllocFilename(xclbin_fnm.data()); // C API, construct xclbin object from filename
+
+  if (xclbinHandle == nullptr)
+    throw std::runtime_error("FAILED_TEST\nxrtXclbinAllocFilename returned nullptr.");
+
+  int xsaSz;
+  if (xrtXclbinGetXSAName(xclbinHandle,NULL,&xsaSz) != C_SUCCESS)
+    throw std::runtime_error("FAILED_TEST\nxrtXclbinGetXSAName returned non-zero error code.");
+  
+  char* xsaName = (char*) malloc(xsaSz*sizeof(char));
```

Comment:
C++ never malloc / free, nor new / delete.  Use std::vector<char> for char buffer arrays, or maybe std::unique_ptr<char[]>

Reply:
fixed